### PR TITLE
Add soundfont tools

### PR DESCRIPTION
**Pull request by schnitzeltony**
* sf2-splitter helps saving memory on embedded machines
* sf2-test checks integrity and outputs useful information

### CODE REVIEW

**Comment by andrewcb** [comment]
The Soundfont tools appear a bit beyond the scope of `iffdigest` (a general-purpose IFF library not specific to any particular application such as SoundFonts). They should probably live in their own repository/project, with `iffdigest` as a dependency.

File: Makefile [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
-CXX	= c++
+#CXX	= c++
 #CXX	= g++
-CXXFLAGS = -g
+# for debug
+#CXXFLAGS += -g -Wall -O0
+CXXFLAGS += -Wall
 
-all:	iffdigest.o
+all:	sf2-split sf2-test
```
Comment (andrewcb) [inline Makefile:8]: Given that this project is an IFF structure parsing library, making the main target be a set of command-line utilities for handling a specific file format doesn't make much sense.

File: sf2.h [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+  std::list<uint16_t> igenIdxList;
+  std::list<uint16_t> shdrIdxList;
+  std::map<uint16_t, uint16_t> shdrIdxMapper; // key:source / val:dest
+  std::map<uint32_t/*start*/, std::map<uint32_t/*end*/, uint32_t/*offset*/>> sampleAreas;
+
+  uint32_t sampleCountTotal;
+};
+
+
+#endif
```
Comment (andrewcb) [inline sf2.h:335]: Good work, btw.

**Comment by schnitzeltony** [comment]
> The Soundfont tools appear a bit beyond the scope of iffdigest (a general-purpose IFF library not specific to any particular application such as SoundFonts). They should probably live in their own repository/project, with iffdigest as a dependency.

I expected this more or less and understand your objection. What do you think about patches 1-4 and 6? They just extend iffdigest slightly. If you agree with them, I could send another pull request removing the soundfont specific patch 5.

Time to mention: Without iffdigest I wouldn't have started this so it's time to say thank you!

FWIW: I added soundfont splitter to my build process [1]. Using single-preset soundfonts reduces memory consumption on RaspberryPi significantly and enables me setups which were not possible before.

[1] https://github.com/schnitzeltony/meta-qt5-extra/commit/a42a532a4a376892d419236061a5ee494996a599

File: iffdigest.cc [hunk 3]
```diff
@@ -53,6 +54,56 @@ IFFChunk::operator=(const IFFChunk& ck)
...
+      subChunkIterator != ck_end(); subChunkIterator++) {
+    (*subChunkIterator).writeData(len, outData, iffFormat);
+  }
+  // ListLen
+  if(ctype == IFF_CHUNK_LIST && outData) {
+    // length itself is not part of length -> '- sizeof(unsigned int)'
+    unsigned int listLen = u32(len - listLenStart - sizeof(unsigned int), iffFormat);
+    memcpy(outData+listLenStart, &listLen, sizeof(listLen));
+  }
+}
```
Comment (andrewcb) [inline iffdigest.cc:105]: Some bounds checking could be good; perhaps with a `maxLen` parameter that the function stops writing if `len` ever exceeds this.
(Also, should the type of an offset be `int` or something like `size_t`?)

**Comment by andrewcb** [comment]
@schnitzeltony: Resubmitting this patch without the SoundFont-specific parts would be great, thank you.

Also, I attached some comments to your `writeData` implementation.

**Comment by schnitzeltony** [comment]
I close this PR to avoid unwanted spam when I rebase my master and send new PR without soundfont and with your suggestions applied.

**Pull request closed without merging**